controllers/operatorscaler_controller.go

```diff
@@ -0,0 +1,364 @@
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "odf.ibm.com/v1alpha1",
+			Kind:       "FlashSystemCluster",
+		},
+	}
+
+	crdNameList = []string{
+		crdNameStorageCluster,
+		crdNameCephCluster,
+		crdNameNoobaa,
+		crdNamePrometheus,
+		crdNameAlertmanager,
+		crdNameFlashSystemCluster,
+	}
+
+	crdNameToKindMapping = map[string]*metav1.PartialObjectMetadata{
+		crdNameStorageCluster:     kindStorageCluster,
+		crdNameCephCluster:        kindCephCluster,
+		crdNameNoobaa:             kindNoobaa,
+		crdNamePrometheus:         kindPrometheus,
+		crdNameAlertmanager:       kindAlertmanager,
+		crdNameFlashSystemCluster: kindFlashSystemCluster,
+	}
+
+	kindToPkgNamesMapping = map[*metav1.PartialObjectMetadata][]string{
+		kindStorageCluster:     {pkgNameOcsOperator},
+		kindCephCluster:        {pkgNameRookCephOperator, pkgNameCephcsiOperator, pkgNameOdfCsiAddonsOperator},
+		kindNoobaa:             {pkgNameMcgOperator},
+		kindPrometheus:         {pkgNameOdfPrometheusOperator},
+		kindAlertmanager:       {pkgNameOdfPrometheusOperator},
+		kindFlashSystemCluster: {pkgNameIbmStorageOdfOperator},
+	}
+
+	createPredicate = predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+)
+
+type OperatorScalerReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	OperatorNamespace string
+	ctx               context.Context
+	log               logr.Logger
+	cache             cache.Cache
+	controller        controller.Controller
+	watching          map[string]bool
+}
+
+//+kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
+//+kubebuilder:rbac:groups=operators.coreos.com,resources=clusterserviceversions,verbs=get;list;update
+
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=noobaa.io,resources=noobaas,verbs=get;list;watch
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=prometheuses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=alertmanagers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=odf.ibm.com,resources=flashsystemclusters,verbs=get;list;watch
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *OperatorScalerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+
+	r.log.Info("starting reconcile")
+
+	err := r.isOdfDependenciesCsvReady()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.addWatches(req.Name)
```

Comment:
This is a named controller which mean you do not know the type of the object on the request. It might be a CRD or a CR and controllers should not behave differently for different kinds on the request. This leaves us with 2 options:
1. separate into 2 different controllers - I don't think we should do that
2. Target all CRDs in addWatches and don't base it on the request - This would be my preference

Reply:
I believe what I'm doing is more efficient. Targeting all crds means getting the list of crds, looking for the ones we're interested in, and then adding watches. We will be running unnecessary code with every reconciliation which can be avoided with the current approach.